packages/admin-panel/src/syncButton/SyncButton.js

```diff
@@ -0,0 +1,129 @@
+/**
+ * Tupaia MediTrak
+ * Copyright (c) 2018 Beyond Essential Systems Pty Ltd
+ */
+
+import React, { useState, useEffect } from 'react';
+import PropTypes from 'prop-types';
+import SyncIcon from '@material-ui/icons/Sync';
+import { Tooltip } from '@material-ui/core';
+import styled, { keyframes } from 'styled-components';
+import { useApi } from '../utilities/ApiProvider';
+import { IconButton } from '../widgets';
+import { makeSubstitutionsInString } from '../utilities';
+
+const STATUSES = {
+  IDLE: 'IDLE',
+  SYNCING: 'SYNCING',
+  ERROR: 'ERROR',
+};
+
+const spin = keyframes`
+    0% { 
+        transform: rotate(360deg); 
+    }
+
+    100% { 
+        transform: rotate(0deg); 
+    }
+`;
+
+const SpinningSyncIcon = styled(SyncIcon)`
+  color: white;
+
+  animation: 3s ${spin};
+  animation-timing-function: linear;
+  animation-iteration-count: infinite;
+`;
+
+const ActiveIconButton = styled(IconButton)`
+  background-color: ${props => props.theme.palette.success.main};
+
+  &.Mui-disabled {
+    background-color: ${props => props.theme.palette.success.main};
+    color: white;
+  }
+`;
+
+const ErrorIconButton = styled(IconButton)`
+  background-color: ${props => props.theme.palette.warning.main};
+  color: ${props => props.theme.palette.common.white};
+
+  &:hover {
+    background-color: ${props => props.theme.palette.warning.dark};
+  }
+
+  &.Mui-disabled {
+    background-color: ${props => props.theme.palette.warning.main};
+    color: white;
+  }
+`;
+
+export const SyncButton = props => {
+  const { actionConfig, original } = props;
+  const api = useApi();
+  const [status, setStatus] = useState(original.sync_status);
+  const [errorMessage, setErrorMessage] = useState(null);
```

Comment:
I try a bit to make its status persisted. Thing is `DataFetchingTable` always triggers a whole refresh, even though I memo the data.. 

And surprisingly in memo the next props data is always empty.. 

I quit since it costed me for a long while debugging... Happy to keep it as it is.

Reply:
One more thing is about the UX on this button. Should it give a green tick if it synced successfully?

Reply:
Thanks so much for looking into it man! Yeah it took a few hours out of my life too 😢 

I actually decided to go with a bit of a hacky solution to persist the latest state outside the component entirely in a map. I was happy to leave it too, but there's a really annoying bug where we lost the button status when opening the logs which can be really confusing to the user (makes it look like the sync is working again when it's really still broken).

Reply:
I use `[shouldComponentUpdate()](https://reactjs.org/docs/react-component.html#shouldcomponentupdate)` in `DataFetchingTable`, set it return false, then it works fine. The thing is we need to work out the right state comparison.

Again, `nextProps.data` is always empty.. but it can be another solution if you can find out why.

Reply:
I feel with such a generic component, working out a standard behaviour for `shouldComponentUpdate()` will be too difficult? What do you think?

Reply:
That's true, the 'always refresh' serves row expandation for survey section, the same as the matrix table in Tupaia. To persist the states correctly, it will be a huge refactor for the related components..

And that explains why in Tupaia, we store states for small components like this in global.🙂

I am satisfied with the hacky solution, I think it is a good way to go. I have raised a question whether this button or something should give a green tick if it synced successfully. Adding a success status could allow us to remove the interval waiting time, and it will be no need for an external state.

Reply:
Raised [a PR](https://github.com/beyondessential/tupaia/pull/4123) to add a shouldComponentUpdate. Comparing `reduxId`, `numberOfPages` and `data` should satisfy the needs?